core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1153,16 +1153,15 @@ private RexNode simplifyIs(RexCall call, RexUnknownAs unknownAs) {
       return rexBuilder.makeLiteral(true);
     }
     if (RexUtil.isLosslessCast(a)) {
-      if (!a.getType().isNullable()) {
-        return rexBuilder.makeLiteral(true);
-      }
-      return rexBuilder.makeCall(SqlStdOperatorTable.IS_NOT_NULL, RexUtil.removeCast(a));
+      a = RexUtil.removeCast(a);
+      // to keep this simplification, we must return IS NOT NULL(a),
+      // even if we cannot do anything else
     }
     if (predicates.pulledUpPredicates.contains(a)) {
       return rexBuilder.makeLiteral(true);
     }
     if (hasCustomNullabilityRules(a.getKind())) {
-      return null;
+      return rexBuilder.makeCall(SqlStdOperatorTable.IS_NOT_NULL, a);
```

Comment:
what is the difference between returning null and creating a new call?
If a is unchanged, perhaps you should return null - less wasteful
Also, please use the makeCall version which preserves the source position.

Reply:
The source pos for the new `IS NULL` / `IS NOT NULL` call is not available yet in the body. The param `a` is the argument to that call, not the call itself. So I would need to add another parameter to that function. I feel that this is out-of-scope for the current ticket. What do you think?

Reply:
ok, I will do it in a separate PR which will do this for more operators.

Reply:
We have built some very nice profiling tools which make use of the source position information. I will add more and more of it if possible.